### PR TITLE
Update callbacks.py to include advance learning scheduler

### PR DESCRIPTION
**Pull request by omarcr**
This advance learning scheduler decrease learning rate according to tolerance.
After decreasing the rate the weights on "filepath" are loaded to allow best fine tuning from the best set of weights.

Most of the code was developed by TIBORG (http://pastebin.com/rJkUdRAz) I only added the load best weights code.

### CODE REVIEW

**Comment by fchollet** [comment]
Docstring is incomplete, API should be improved. Use case unclear. And it seems you included changes that are not related to the PR.

**Comment by omarcr** [comment]
It is an old version keras considered. 
The only function modified is the  is the learning scheduler. This function is widely used in all deep learning models for finetuning, something that Keras currently lacks on. The learning scheduler available is very basic and doesn't decrease the learning rate at a proper rate.

File: keras/callbacks.py [hunk 1]
```diff
@@ -516,3 +516,78 @@ def on_epoch_end(self, epoch, logs={}):
...
+        if self.monitor_op(current, self.best):
+            self.best = current
+            self.wait = 0
+        else:
+            if self.wait >= self.patience:
+                if self.verbose > 0:
+                    print('\nEpoch %05d: reducing learning rate' % (epoch))
+                    assert hasattr(self.model.optimizer, 'lr'), \
+                        'Optimizer must have a "lr" attribute.'
+                    current_lr = K.get_value(self.model.optimizer.lr)
```
Comment (esube) [inline keras/callbacks.py]: Code below this line should be on same indentation as if self.wait >= self.patience
Otherwise, if verbosity is turned off, the current implementation doesn't do the actual learning rate scheduling.

File: keras/callbacks.py [hunk 2]
```diff
@@ -516,3 +516,78 @@ def on_epoch_end(self, epoch, logs={}):
...
+                          (self.monitor), RuntimeWarning)
+ 
+        if self.monitor_op(current, self.best):
+            self.best = current
+            self.wait = 0
+        else:
+            if self.wait >= self.patience:
+                if self.verbose > 0:
+                    print('\nEpoch %05d: reducing learning rate' % (epoch))
+                    assert hasattr(self.model.optimizer, 'lr'), \
```
Comment (esube) [inline keras/callbacks.py]: Code below this line should be on same indentation as if self.wait >= self.patience
Otherwise, if verbosity is turned off, the current implementation doesn't do the actual learning rate scheduling.

File: keras/callbacks.py [hunk 3]
```diff
@@ -516,3 +516,78 @@ def on_epoch_end(self, epoch, logs={}):
...
+                        'Optimizer must have a "lr" attribute.'
```
Comment (esube) [inline keras/callbacks.py]: Code below this line should be on same indentation as if self.wait >= self.patience
Otherwise, if verbosity is turned off, the current implementation doesn't do the actual learning rate scheduling.

File: keras/callbacks.py [hunk 4]
```diff
@@ -471,11 +471,11 @@ def _set_model(self, model):
             layers = self.model.layers
             for layer in layers:
                 if hasattr(layer, 'W'):
-                    tf.histogram_summary('{}_W'.format(layer.name), layer.W)
+                    tf.histogram_summary('{}_W'.format(layer), layer.W)
```
Comment (esube) [inline keras/callbacks.py]: I am not sure how this changes are related to this PR?

**Comment by esube** [comment]
This is essentially a step learning rate scheduler. It drops the learning rate when the monitored param doesn't drop for patience amount of epochs. This particular PR deosn't seem to stop the learning and rather keep reducing the learning rate by the decay factor every patience number of times. Plus, I am not sure how fine-tuning and step learning rate scheduling are tied here? step learning rate schedule can work for any case, either fine-tuning or training from scratch.

But, I am not sure, if it is a good idea to keep adding these schedulers in to the main callback. There is also a version which is popular in training resnets which drops the learning rate at specified discrete epochs specified by user and there is also another one with restarts and I am sure there will be more complicated scheduling coming in the future. So, it is not feasible to implement all of them separately and add them to keras. 

I think, it will rather be wise to improve the current LearningRateScheduler and let the user implement the actual callback functions.

**Comment by omarcr** [comment]
This is a essential tool for fine tuning. If we don't put it into the main callbacks users will do it manually which is a pain or they have to implement their own, which will be a waste of time considering that we can offer them one function straight out of the box.

**Comment by fchollet** [comment]
I'm all for including better learning rate scheduling support. But this PR should be heavily revised. Please resubmit when you have made appropriate changes.
